File: Android/app/src/main/java/jp/co/riso/smartdeviceapp/view/printersearchsettings/SnmpCommunityNameDisplayText.java
/*
 * Copyright (c) 2015 Ricoh Company, Ltd. All rights reserved.
 *
 * SnmpCommunityNameDisplayText.java
 * SmartDeviceApp
 * Created by: a-LINK Group
 */

package jp.co.riso.smartdeviceapp.view.printersearchsettings;

import android.content.Context;
import android.os.Parcel;
import android.os.Parcelable;
import android.preference.PreferenceManager;
import android.util.AttributeSet;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;

import jp.co.riso.smartdeviceapp.AppConstants;
import jp.co.riso.smartprint.R;

public class SnmpCommunityNameDisplayText extends LinearLayout {

    private String communityName = AppConstants.PREF_DEFAULT_SNMP_COMMUNITY_NAME;

    public SnmpCommunityNameDisplayText(Context context) {
        super(context);
        init(context);
    }

    public SnmpCommunityNameDisplayText(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }

    public SnmpCommunityNameDisplayText(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context);
    }

    public SnmpCommunityNameDisplayText(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        init(context);
    }

    private void init(Context context) {
        View view = View.inflate(context, R.layout.printersearchsettings_community_name_display, this);

        this.communityName = PreferenceManager.getDefaultSharedPreferences(context).getString(AppConstants.PREF_KEY_SNMP_COMMUNITY_NAME, AppConstants.PREF_DEFAULT_SNMP_COMMUNITY_NAME);

        TextView textView = (TextView) view.findViewById(R.id.snmpCommunityName);
        textView.setText(this.communityName);
    }

    @Override
    public Parcelable onSaveInstanceState() {
        Parcelable superState = super.onSaveInstanceState();
        SavedState savedState = new SavedState(superState);

        savedState.communityName = this.communityName;

        return savedState;
    }

    @Override
    public void onRestoreInstanceState(Parcelable state) {
        if(!(state instanceof SavedState)) {
            super.onRestoreInstanceState(state);
            return;
        }

        SavedState savedState = (SavedState)state;
        super.onRestoreInstanceState(savedState.getSuperState());

        this.communityName = savedState.communityName;
    }

    static class SavedState extends BaseSavedState {
        String communityName;

        SavedState(Parcelable superState) {
            super(superState);
        }

        private SavedState(Parcel in) {
            super(in);
            this.communityName = in.readString();
        }

        @Override
        public void writeToParcel(Parcel out, int flags) {
            super.writeToParcel(out, flags);
            out.writeString(this.communityName);
        }

        //required field that makes Parcelables from a Parcel
        public static final Parcelable.Creator<SavedState> CREATOR =
                new Parcelable.Creator<SavedState>() {
                    public SavedState createFromParcel(Parcel in) {
                        return new SavedState(in);
                    }
                    public SavedState[] newArray(int size) {
                        return new SavedState[size];
                    }
                };
    }
}

File: Android/app/src/main/java/jp/co/riso/smartdeviceapp/view/printersearchsettings/SnmpCommunityNameDisplayText.kt
/*
 * Copyright (c) 2015 Ricoh Company, Ltd. All rights reserved.
 *
 * SnmpCommunityNameDisplayText.java
 * SmartDeviceApp
 * Created by: a-LINK Group
 */
package jp.co.riso.smartdeviceapp.view.printersearchsettings

import android.content.Context
import android.widget.LinearLayout
import jp.co.riso.smartdeviceapp.AppConstants
import jp.co.riso.smartprint.R
import android.widget.TextView
import android.os.Parcelable
import android.os.Parcel
import android.os.Parcelable.Creator
import android.util.AttributeSet
import androidx.preference.PreferenceManager

class SnmpCommunityNameDisplayText : LinearLayout {
    private var _communityName: String? = AppConstants.PREF_DEFAULT_SNMP_COMMUNITY_NAME

    constructor(context: Context) : super(context) {
        init(context)
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        init(context)
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    ) {
        init(context)
    }

    private fun init(context: Context) {
        val view = inflate(context, R.layout.printersearchsettings_community_name_display, this)
        _communityName = PreferenceManager.getDefaultSharedPreferences(context).getString(
            AppConstants.PREF_KEY_SNMP_COMMUNITY_NAME,
            AppConstants.PREF_DEFAULT_SNMP_COMMUNITY_NAME
        )
        val textView = view.findViewById<TextView>(R.id.snmpCommunityName)
        textView.text = _communityName
    }

    public override fun onSaveInstanceState(): Parcelable {
        val superState = super.onSaveInstanceState()
        val savedState = SavedState(superState)
        savedState.communityName = _communityName
        return savedState
    }

    public override fun onRestoreInstanceState(state: Parcelable) {
        if (state !is SavedState) {
            super.onRestoreInstanceState(state)
            return
        }
        super.onRestoreInstanceState(state.superState)
        _communityName = state.communityName
    }

    internal class SavedState : BaseSavedState {
        var communityName: String? = null

        constructor(superState: Parcelable?) : super(superState)
        private constructor(`in`: Parcel) : super(`in`) {
            communityName = `in`.readString()
        }

        override fun writeToParcel(out: Parcel, flags: Int) {
            super.writeToParcel(out, flags)
            out.writeString(communityName)
        }

        companion object {
            //required field that makes Parcelables from a Parcel
            @JvmField
            val CREATOR: Creator<SavedState?> = object : Creator<SavedState?> {
                override fun createFromParcel(`in`: Parcel): SavedState {
                    return SavedState(`in`)
                }

                override fun newArray(size: Int): Array<SavedState?> {
                    return arrayOfNulls(size)
                }
            }
        }
    }
}